### DISCUSSION TITLE
Optimizing Dielectric Materials

### DISCUSSION THREAD

**Discussion by sapo17** [mitsuba-renderer/mitsuba3#379] [Q&A] [answered]
Hello,

I am quite new to Mitsuba, so apologies upfront if I'm getting things wrong.

So, I was naively trying to optimize the 'eta' value of a smooth dielectric material, which after some steps* I realized that its not differentiable .

Interestingly, I am able to optimize mirror materials eta value (although not with detached samplers, like the 'prb' integrator) - which is great.

So, my question is:
What is actually the reason for it? Are there any literature findings that you may refer me to?
Are there maybe some workarounds that one can do?
I saw on the documentation that the 'eta' values of thin dielectric and rough dielectric materials are differentiable (not tested yet)
I also saw on the documentation the following regarding rough dielectric:

> This plugin is essentially the roughened equivalent of the (smooth) plugin [dielectric](https://mitsuba.readthedocs.io/en/stable/src/generated/plugins_bsdfs.html#bsdf-dielectric).

Would that be actually a possible way to imitate soft dielectric materials (using lower alpha values)?

Apologies if I'm causing headaches. I am writing my bachelor thesis regarding optimizing translucent materials for lights. Mitsuba is helping me immensly (thank you).

Best,
Can - University of Vienna

*Unfortunately, I was unware of the 'Flag' that each scene parameter got, so I tried to find workarounds casting mi.Float to float and vice versa. Nevertheless, I didn't managed to work it out since either float type was not differentiable (i.e. exception) or casting original float value into mi.Float and updating scene parameters caused exceptions

**Comment by Speierers** [comment, marked as answer]
Hi @sapo17,

As explained in [this paper Fig 5](https://rgl.epfl.ch/publications/Zeltner2021MonteCarlo) (and others), differentiating parameters of a perfectly specular BSDF (such as `eta`) will require handling discontinuities. For that, you could try to use the `prb_reparam` integrator instead, although I am not sure we have ever tested this for optimizing the `eta` parameter.

**Reply by maxfrei750** [reply]
If I may, I'd like to ask a follow-up question: What exactly do you mean with 

> perfectly specular BSDF

only materials with `spec_trans=1` or all materials with `spec_trans>0`?

**Reply by ziyi-zhang** [reply]
Hi @maxfrei750 
I believe the term `perfectly specular BSDF` is referring to the materials whose sampling distribution is a Dirac delta function. So no matter it's reflection or refraction, so long as the BSDF is smooth (dielectric or conductor), it would be non-trivial to optimize `eta`.

I also doubt the correctness of using reparam-based integrators here. Reparam method adds some perturbation to the outgoing direction to account for the discontinuous color change. Degenerated BSDFs violate the assumption of this method. 

For now, one workaround in Mitsuba is to use a rough dielectric BSDF (with roughness set to be a small number) to mimic a smooth dielectric. I hope that prb-reparam would produce correct gradients in this case.

**Reply by maxfrei750** [reply]
Hi @ziyi-zhang, thanks for elaborating. That definitively helped my understanding.
